bittensor/_dataset/dataset_impl.py

```diff
@@ -1,626 +1,800 @@
+                Include self.data_dir as the prefix.
+                    - if True, ths meants shortens the batch and 
+                    specializes it to be with respect to the dataset's 
+                    root path which is in ./bittensor/dataset
             
-            if self.save_dataset and self.dataset_hashes[file_meta['Folder']]['Size'] < self.backup_dataset_cap_size:
-                self.save_hash( file_meta, text )
-                self.dataset_hashes[file_meta['Folder']]['Size'] += file_meta['Size']
+        Returns: path (str)
+            Path of the saved JSON
+        """
+        
+        
+        if include_root:
+            path = os.path.join(self.data_dir, path)
+
+
+        # ensure extension
+        dir_path = os.path.dirname(path)
+        if path[-len('.json'):] != '.json':
+            path += '.json'
+
+        # ensure dictionary
+        if not os.path.isdir(dir_path):
+            os.makedirs(dir_path)
+
+
+        # load default if file does not exist
+        try:
+            with open(path, 'r') as f:
+                obj = json.load(f)
+        except FileNotFoundError:
+            obj = default
+
+
+        if isinstance(obj, str):
+            obj = json.loads(obj)
+
+
+        return obj
+
+    def load_json(self, loop:'asyncio.loop'=None, *args,**kwargs) -> Union[list, dict]:
+        '''
+        Sync verson of async_save_json
+
+        Args
+            loop (asyncio.loop):
+                The asyncio loop to be past, otherwise self.loop
+
+        Returns (dict, list) 
+
+        '''
+        return self.async_run(job=self.async_load_json(*args,**kwargs), loop=loop)
+
+    
+    async def build_dataset(self, dataset=None, num_folders=10, num_samples=100, save=False, load=True, loop=None):
+
+        folder_hashes = (await self.get_folder_hashes(self.dataset2hash[dataset]))[:num_folders]
+        random.shuffle(folder_hashes)
+
+        loaded_text_hashes, new_text_hashes = [], []
+        if load:
+            loaded_text_hashes =  self.load_json(path=f'{dataset}/hashes', default=[], loop=loop)
+            if len(loaded_text_hashes)>num_samples:
+                return loaded_text_hashes[:num_samples]
+
+        for f in folder_hashes:
             
-        else:
-            logger.warning("Failed to get text".ljust(20) + "<blue>{}</blue>".format(file_meta['Name']))
-            self.IPFS_fails += 1
+            self.total = 0
+            folder_text_hashes = await self.get_text_hashes(f)
+            new_text_hashes += folder_text_hashes 
             
-        return text 
 
-    def get_dataset(self , file_meta):
-        r""" Either load a dataset, which is a list of hashes, from disk or download it from IPFS
-        Args:
-            file_meta (dict of str: int):
-                Specify the details of the dataset in the format of {'Name': , 'Hash':}.
+            if (len(new_text_hashes) + len(loaded_text_hashes)) >num_samples:
+                break
```

Comment:
Let's leave a message here too to explain why we broke out.

Reply:
This is now depracted as i use a dictionary to better track new hashes.